hazelcast/src/main/java/com/hazelcast/sql/impl/type/converter/FloatConverter.java

```diff
@@ -38,27 +40,58 @@ private FloatConverter() {
 
     @Override
     public byte asTinyint(Object val) {
-        return (byte) cast(val);
+        float casted = cast(val);
+        byte converted = (byte) casted;
+
+        if (converted != (int) casted || !Float.isFinite(casted)) {
+            throw cannotConvert(QueryDataTypeFamily.TINYINT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public short asSmallint(Object val) {
-        return (short) cast(val);
+        float casted = cast(val);
+        short converted = (short) casted;
+
+        if (converted != (int) casted || !Float.isFinite(casted)) {
+            throw cannotConvert(QueryDataTypeFamily.SMALLINT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public int asInt(Object val) {
-        return (int) cast(val);
+        float casted = cast(val);
+        int converted = (int) casted;
+
+        if (converted != (long) casted || !Float.isFinite(casted)) {
+            throw cannotConvert(QueryDataTypeFamily.INT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public long asBigint(Object val) {
-        return (long) cast(val);
+        float casted = cast(val);
+        float truncated = (float) (casted > 0.0 ? Math.floor(casted) : Math.ceil(casted));
+        long converted = (long) truncated;
+
+        // No checks for NaNs and infinities are needed: NaNs are zeros and
+        // infinities are Long.MAX/MIN_VALUE when converted to long.
+        if ((float) converted != truncated) {
+            throw cannotConvert(QueryDataTypeFamily.BIGINT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public BigDecimal asDecimal(Object val) {
-        return BigDecimal.valueOf(cast(val));
+        return new BigDecimal(Float.toString(cast(val)), DECIMAL_MATH_CONTEXT);
```

Comment:
The `(String val, MathContext mc)` constructor is really heavy. Is there a reason why we cannot use `(Double val, MathContext mc)` here? 
I look at `asDouble` implementation, and there are no additional checks there.

Reply:
To my understanding, what users usually want when they cast, for instance, `0.3d` to decimal is a decimal value that is literally equal to `0.3` and not `0.299999999999999988897769753748434595763683319091796875` which is the actual decimal fraction encoded by `0.3d`. Probably we should introduce two different conversion methods, one to convert literals to human-readable decimals and one to convert some intermediate results to unhuman decimals, or just always use the latter ones.

Not quite sure about which `asDouble` and checks you are talking about.

Reply:
Converters are not concerned with literals AFAIK. Literal conversions are parsing-time decisions, while the converters are used during query execution. So I am not sure that we should resort to intermediate string conversions here - it will negatively affect the performance.

Reply:
fixed